File: NativeFlutter/NativeFlutter/Flutter/UsersBridgeChannel.swift
//
//  UsersBridgeChannel.swift
//  NativeFlutter
//
//  MethodChannel — request/response between Flutter and Swift.
//
//  Channel name: "com.huh.nativeflutter/users"
//
//  Methods Flutter can call:
//    - getUsers()                              -> [Map<String,dynamic>]
//    - createUser({name, email})               -> Map<String,dynamic>
//    - deleteUser({id})                        -> bool
//    - closeFlutter()                          -> null   (asks iOS to pop the Flutter screen)
//
//  Methods Swift calls on Flutter:
//    - presentRoute(String)                    // "users" | "create"
//

import Foundation

extension Notification.Name {
    /// Posted when Flutter asks the host (via `closeFlutter`) to
    /// dismiss the embedded Flutter screen. SwiftUI listens for it
    /// and pops the navigation destination.
    static let flutterRequestedClose = Notification.Name("com.huh.nativeflutter.flutterRequestedClose")
}

#if canImport(Flutter)
import Flutter

@MainActor
final class UsersBridgeChannel {
    static let channelName = "com.huh.nativeflutter/users"

    private let channel: FlutterMethodChannel
    private let store: UsersStore

    init(engine: FlutterEngine, store: UsersStore) {
        self.channel = FlutterMethodChannel(
            name: Self.channelName,
            binaryMessenger: engine.binaryMessenger
        )
        self.store = store

        channel.setMethodCallHandler { [weak self] call, result in
            Task { @MainActor in
                self?.handle(call: call, result: result)
            }
        }
    }

    /// Swift → Flutter. Used by `FlutterEngineManager.presentRoute(_:)`
    /// to tell the Flutter side which screen to land on.
    func invoke(method: String, arguments: Any?) {
        channel.invokeMethod(method, arguments: arguments)
    }

    private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getUsers":
            result(store.bridgeDictionaries())

        case "createUser":
            guard
                let args = call.arguments as? [String: Any],
                let name = (args["name"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
                let email = (args["email"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
                !name.isEmpty, !email.isEmpty
            else {
                result(FlutterError(
                    code: "invalid_arguments",
                    message: "createUser expects { name: String, email: String }",
                    details: nil
                ))
                return
            }
            let user = store.add(name: name, email: email, source: .flutter)
            result(user.toBridgeDictionary())

        case "deleteUser":
            guard
                let args = call.arguments as? [String: Any],
                let id = args["id"] as? String
            else {
                result(FlutterError(
                    code: "invalid_arguments",
                    message: "deleteUser expects { id: String }",
                    details: nil
                ))
                return
            }
            store.delete(id: id)
            result(true)

        case "closeFlutter":
            NotificationCenter.default.post(name: .flutterRequestedClose, object: nil)
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

#endif

File: NativeAndroid/app/src/main/java/com/example/androidnative/flutter/UsersBridgeChannel.kt
package com.example.androidnative.flutter

import com.example.androidnative.models.User
import com.example.androidnative.store.UsersStore
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class UsersBridgeChannel(
    messenger: BinaryMessenger,
    private val store: UsersStore,
    private val onCloseFlutter: () -> Unit,
) : MethodChannel.MethodCallHandler {

    private val channel = MethodChannel(messenger, CHANNEL_NAME)

    init {
        channel.setMethodCallHandler(this)
    }

    fun invokeMethod(method: String, arguments: Any?) {
        channel.invokeMethod(method, arguments)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getUsers" -> result.success(store.bridgeMaps())

            "createUser" -> {
                val name = (call.argument<String>("name") ?: "").trim()
                val email = (call.argument<String>("email") ?: "").trim()
                if (name.isEmpty() || email.isEmpty()) {
                    result.error("invalid_arguments", "createUser expects name and email", null)
                    return
                }
                val user = store.add(name = name, email = email, source = User.Source.FLUTTER)
                result.success(user.toBridgeMap())
            }

            "deleteUser" -> {
                val id = call.argument<String>("id")
                if (id.isNullOrEmpty()) {
                    result.error("invalid_arguments", "deleteUser expects id", null)
                    return
                }
                store.delete(id)
                result.success(true)
            }

            "closeFlutter" -> {
                onCloseFlutter()
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    fun destroy() {
        channel.setMethodCallHandler(null)
    }

    companion object {
        const val CHANNEL_NAME = "com.huh.nativeflutter/users"
    }
}
